pkg/client/cli/output/output.go

```diff
@@ -1,162 +1,212 @@
-	if o == nil {
-		return os.Stdout, os.Stderr
+// Out returns an io.Writer that writes to the OutOrStdout of the current *cobra.Command, or
+// if no command is active, to the os.Stdout. If formatted output is requested, the output
+// will be delayed until Execute is called.
+func Out(ctx context.Context) io.Writer {
+	if cmd, ok := ctx.Value(key{}).(*cobra.Command); ok {
+		return cmd.OutOrStdout()
 	}
-
-	return o.stdout, o.stderr
+	return os.Stdout
 }
 
-func SetJSONStdout(ctx context.Context) {
-	o, _ := ctx.Value(key{}).(*output)
-	if o == nil {
-		return
+// Err returns an io.Writer that writes to the ErrOrStderr of the current *cobra.Command, or
+// if no command is active, to the os.Stderr. If formatted output is requested, the output
+// will be delayed until Execute is called.
+func Err(ctx context.Context) io.Writer {
+	if cmd, ok := ctx.Value(key{}).(*cobra.Command); ok {
+		return cmd.ErrOrStderr()
 	}
-
-	o.stdoutIsJSON = true
+	return os.Stderr
 }
 
-func SetJSONStderr(ctx context.Context) {
-	o, _ := ctx.Value(key{}).(*output)
-	if o != nil {
-		return
+// Info is similar to Out, but if formatted output is requested, the output will be discarded.
+//
+// Info is primarily intended for messages that are not directly related to the command that
+// executes, such as messages about starting up daemons or being connected to a context.
+func Info(ctx context.Context) io.Writer {
+	if cmd, ok := ctx.Value(key{}).(*cobra.Command); ok {
+		if _, ok := cmd.OutOrStdout().(*output); ok {
+			return io.Discard
+		}
+		return cmd.OutOrStdout()
 	}
-
-	o.stdoutIsJSON = true
+	return os.Stdout
 }
 
-type (
-	key    struct{}
-	output struct {
-		cmd string
-
-		stdoutBuf strings.Builder
-		stderrBuf strings.Builder
-
-		stdoutIsJSON bool
-		stderrIsJSON bool
-
-		jsonEncoder *json.Encoder
-		stdout      io.Writer
-		stderr      io.Writer
-
-		originalStdout io.Writer
-		originalStderr io.Writer
-	}
-)
-
-func (o *output) runE(f func(cmd *cobra.Command, args []string) error) func(cmd *cobra.Command, args []string) error {
-	return func(cmd *cobra.Command, args []string) error {
-		if !WantsJSONOutput(cmd.Flags()) {
-			return f(cmd, args)
+// Object sets the object to be marshalled and printed on stdout when formatted output
+// is requested using the `--output=<fmt>` flag. Otherwise, this function does nothing.
+//
+// If override is set to true, then the formatted output will consist solely of the given
+// object. There will be no "cmd", "stdout", or "stderr" tags.
+//
+// The function will panic if data already has been written to the stdout of the command
+// or if an Object already has been called.
+func Object(ctx context.Context, obj any, override bool) {
+	if cmd, ok := ctx.Value(key{}).(*cobra.Command); ok {
+		if o, ok := cmd.OutOrStdout().(*output); ok {
+			if o.Len() > 0 {
+				panic("output.Object cannot be used together with output.Out")
+			}
+			if o.obj != nil {
+				panic("output.Object can only be used once")
```

Comment:
in the case of a panic, we might still want to bubble up the object if its a `fmt.Stringer`, especially if we're choosing to not use json output on a panic.

Reply:
I'm not sure I see the point in doing that. A panic here means that a developer made a coding error. It's not intended for the user.